core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java

```diff
@@ -0,0 +1,150 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.reflect.ClassPath;
+import com.google.common.reflect.ClassPath.ClassInfo;
+import io.trino.spi.connector.ConnectorContext;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.Sets.difference;
+import static java.lang.ClassLoader.getPlatformClassLoader;
+import static java.lang.ClassLoader.getSystemClassLoader;
+import static java.lang.reflect.Modifier.isPublic;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestSpiBackwardCompatibility
+{
+    private static final Map<String, Set<String>> BACKWARD_INCOMPATIBLE_CHANGES = ImmutableMap.<String, Set<String>>builder()
+            // When updating this map, please try to remove backward incompatible changes for old versions.
+            // Also consider mentioning backward incompatible changes in release notes.
+            // We try to be backward compatible with at least the last released version.
+            .put("376", ImmutableSet.of(
+                    "Constructor: public io.trino.spi.memory.MemoryPoolInfo(long,long,long,java.util.Map<io.trino.spi.QueryId, java.lang.Long>,java.util.Map<io.trino.spi.QueryId, java.util.List<io.trino.spi.memory.MemoryAllocation>>,java.util.Map<io.trino.spi.QueryId, java.lang.Long>)"))
+            .buildOrThrow();
+
+    @Test
+    public void testSpiSingleVersionBackwardCompatibility()
+            throws Exception
+    {
+        assertThat(getCurrentSpi()).containsAll(difference(getPreviousSpi(), getBackwardIncompatibleChanges()));
+    }
+
+    @Test
+    public void testBackwardIncompatibleEntitiesAreInPreviousSpi()
+            throws Exception
+    {
+        assertThat(getPreviousSpi()).containsAll(getBackwardIncompatibleChanges());
+    }
+
+    private static Set<String> getBackwardIncompatibleChanges()
+    {
+        String spiVersion = new ConnectorContext() {}.getSpiVersion();
+        int currentVersion = Integer.parseInt(spiVersion.replace("-SNAPSHOT", ""));
+        String previousVersion = Integer.toString(currentVersion - 1);
+        return BACKWARD_INCOMPATIBLE_CHANGES.getOrDefault(previousVersion, ImmutableSet.of());
```

Comment:
`BACKWARD_INCOMPATIBLE_CHANGES.get...(currentVersion)` -- "the backwards incompatible changes introduced in version (current) are ..."